pkg/apis/storage/types.go

```diff
@@ -80,3 +80,95 @@ type StorageClassList struct {
 	// Items is the list of StorageClasses
 	Items []StorageClass
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Captures the intent to attach or detach the specified volume to/from
+// the specified node.
+//
+// VolumeAttachment objects are non-namespaced.
+type VolumeAttachment struct {
+	metav1.TypeMeta
+
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Specification of the desired attach/detach volume behavior.
+	// Populated by the Kubernetes system.
+	Spec VolumeAttachmentSpec
+
+	// Status of the VolumeAttachment request.
+	// Populated by the entity completing the attach or detach
+	// operation, i.e. the external-attacher.
+	Status VolumeAttachmentStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeAttachmentList is a collection of VolumeAttachment objects.
+type VolumeAttachmentList struct {
+	metav1.TypeMeta
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is the list of VolumeAttachments
+	Items []VolumeAttachment
+}
+
+// The specification of a VolumeAttachment request.
+type VolumeAttachmentSpec struct {
+	// The name of the external-attacher that MUST handle
+	// this request.
+	Attacher string
+
+	// The volume to attach.
+	Volume api.VolumeSource
+
+	// The node that the volume should be attached to.
+	NodeName string
+}
+
+// The status of a VolumeAttachment request.
+type VolumeAttachmentStatus struct {
+	// Indicates the volume is successfully attached.
+	// This field must only be set by the entity completing the attach
+	// operation, i.e. the external-attacher.
+	IsAttached bool
```

Comment:
a couple things to consider from the [API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md):
> * Think twice about bool fields. Many ideas start as boolean but eventually trend towards a small set of mutually exclusive options. Plan for future expansions by describing the policy options explicitly as a string type alias (e.g. TerminationMessagePolicy).
> * The name of a field expressing a boolean property called 'fooable' should be called Fooable, not IsFooable.

Reply:
Also, it seems like we'd need to be able to tell from status whether attach had begun.

Since attach is not instant, there could be a period of time during 

1. Kubernetes creates VolumeAttachment
2. Attacher observes VolumeAttachment, starts attaching
3. PV/VolumeAttachment objects are vulnerable to deletion/finalization, since status doesn't indicate the volume is attached or is attaching
4. Attacher completes attaching, tries to update status to indicate attached state, object is gone

It seems like it should be possible to tell where we are in the `detached -> attaching -> attached -> detaching -> detached` lifecycle by looking at status (and making the attacher do a successful status update from `detached -> attaching` before starting the attach operation prevents races trying to attach a volume that is getting deleted)

Reply:
The idea is that we do not need "attaching" state, simple presence of the object with `(Is)Attached==false` is enough. The same for "detaching", presence of `deletetionTimestamp` is enough.

> `detached` -> `attaching`
This is not possible. VolumeAttachment with `deletetionTimestamp` must be deleted first in API server and Kubernetes must create a new instance. Or is it allowed to cancel deletetionTimestamp? It looks odd.

Reply:
We did not want to introduce state machines with detached -> attaching -> attached -> detaching -> detached because it's not possible to add a new state there later.

Reply:
>> detached -> attaching
>
> This is not possible

Maybe I used the wrong term. Is `unattached -> attaching -> attached -> detaching -> detached` more correct? Don't we need to represent the state where the attacher has observed the request for the attachment and is currently working to attach (so the VolumeAttachment object should not be removed), but it has not yet attached (so the volume is not ready to mount)?

Reply:
I don't think we need to capture the state that attacher has started attaching. A/D controller with time out eventually and send some event to pod. There is a glitch though, there is no way how to tell if external attacher does not run or it's just slow.